Add Terraform Kubernetes module comparison blog post (#28369)
kubernetes/website: content/en/blog/_posts/2021-06-14-a-better-way-to-provision-kubernetes-resources-using-terraform.md

---
layout: blog
title: 'A Better Way to Provision Kubernetes Resources Using Terraform'
date: 2021-06-14
slug: a-better-way-to-provision-kubernetes-resources-using-terraform
---

**Author:** Philipp Strube (Kubestack)

> After months of work and thousands of lines of Go, HCL and YAML I am very excited to share the release of the new [Kustomize based Terraform modules for Kubernetes](https://www.kubestack.com/catalog).

Reviewer: ```suggestion
After months of work and thousands of lines of Go, HCL, and YAML, I am very excited to share the release of the new [Kustomize-based Terraform modules for Kubernetes](https://www.kubestack.com/catalog).

```

Reviewer: s/ based/-based/ (multiple instances)

The modules use the [Kustomization provider](https://registry.terraform.io/providers/kbst/kustomization/latest) I maintain as part of my work on the [Kubestack framework](https://www.kubestack.com/).
Kubestack is an open-source Terraform framework and strives to deliver the best developer experience for teams working with Terraform and Kubernetes.

> I believe the new modules are a better way to provision Kubernetes resources using Terraform, because they bring three distinct benefits over the existing alternatives:

Reviewer: ```suggestion
I believe the new modules are a better way to provision Kubernetes resources using Terraform because they bring three distinct benefits over the existing alternatives:

```

> 1. Like Kustomize, the upstream YAML is left untouched, meaning upstream updates are low maintenance effort

Reviewer: ```suggestion
1. Like Kustomize, the upstream YAML is left untouched, meaning upstream updates are a low-maintenance effort.

```

> 2. By defining the Kustomize overlay in HCL, all Kubernetes resources are fully customizeable using values from Terraform

Reviewer: ```suggestion
2. By defining the Kustomize overlay in HCL, all Kubernetes resources are fully customizable using values from Terraform.

```

> 3. Each Kubernetes resources is tracked individually in Terraform state, so diffs and plans show the changes to the actual Kubernetes resources

Reviewer: ```suggestion
3. Each Kubernetes resource is tracked individually in Terraform state, so diffs and plans show the changes to the actual Kubernetes resources.

```

> To make these benefits less abstract, lets compare the [Nginx ingress module](https://www.kubestack.com/catalog/nginx) with [one using the Helm provider](https://registry.terraform.io/modules/terraform-module/release/helm/latest) to provision [Nginx ingress](https://github.com/kubernetes/ingress-nginx).

Reviewer: s/lets/let's/ (3 instances)

The Terraform configuration for both examples is available in [this repository](https://github.com/kbst/terraform-helm-vs-kustomize). Lets take a look at the Helm module first.

## The Helm based module

Usage of the module is straightforward. First, we configure the Kubernetes and Helm providers.

```hcl
provider "kubernetes" {
  config_path    = "~/.kube/config"
}

provider "helm" {
  kubernetes {
    config_path = "~/.kube/config"
  }
}
```

Then we define a `kubernetes_namespace` and call the `release/helm` module.

```hcl
resource "kubernetes_namespace" "nginx_ingress" {
  metadata {
    name = "ingress-nginx"
  }
}

module "nginx_ingress" {
  source  = "terraform-module/release/helm"
  version = "2.6.10"

  namespace  = kubernetes_namespace.nginx_ingress.metadata.0.name
  repository =  "https://kubernetes.github.io/ingress-nginx"

  app = {
    name          = "ingress-nginx"
    version       = "3.33.0"
    chart         = "ingress-nginx"
    force_update  = true
    wait          = false
    recreate_pods = false
    deploy        = 1
  }

  set = [
    {
      name = "replicaCount"
      value = 2
    }
  ]
}
```

> If we now run a `terraform plan` for this configuration, we will see the resources to be created.

Reviewer: To subtly increase the power of your writing, look for "will" and try to replace it with the present-tense.
> If we now run a `terraform plan` for this configuration, we will see the resources to be created.
becomes
> If we now run a `terraform plan` for this configuration, we see the resources to be created.

      + reuse_values               = false
      + skip_crds                  = false
      + status                     = "deployed"
      + timeout                    = 300
      + values                     = []
      + verify                     = false
      + version                    = "3.33.0"
      + wait                       = false
      + wait_for_jobs              = false

      + set {
          + name  = "replicaCount"
          + value = "2"
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

And this is the key issue with how Helm is integrated into the Terraform workflow.
The plan does not tell me what Kubernetes resources will be created for the Nginx ingress controller.
And neither are the Kubernetes resources tracked in Terraform state, as shown by the apply output.

```hcl
kubernetes_namespace.nginx_ingress: Creating...
kubernetes_namespace.nginx_ingress: Creation complete after 0s [id=ingress-nginx]
module.nginx_ingress.helm_release.this[0]: Creating...
module.nginx_ingress.helm_release.this[0]: Creation complete after 6s [id=ingress-nginx]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

> Similarly, if planning a change there's again no way to tell what the changes to the Kubernetes resources will be.

Reviewer: ```suggestion
Similarly, if planning a change, there's again no way to tell what the changes to the Kubernetes resources will be.

```

So if we increase the `replicaCount` value of the Helm chart, the `terraform plan` will merely show the change to the `helm_release` resource.

```hcl
set = [
  {
    name = "replicaCount"
    value = 2
  }
]
```

What will the changes to the Kubernetes resources be?

> And more importantly, is it a simple in-place update or does it require a destroy and recreate?

Reviewer: ```suggestion
And more importantly, is it a simple in-place update, or does it require a destroy-and-recreate?

```

        name                       = "ingress-nginx"
        # (27 unchanged attributes hidden)

      - set {
          - name  = "replicaCount" -> null
          - value = "2" -> null
        }
      + set {
          + name  = "replicaCount"
          + value = "3"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

## The Kustomize based module

Now, lets take a look at the same steps for the Kustomize based module.
Usage is similar.
First we require the `kbst/kustomization` provider and configure it.

```hcl
terraform {
  required_providers {
    kustomization = {
      source = "kbst/kustomization"
    }
  }
}

provider "kustomization" {
  kubeconfig_path = "~/.kube/config"
}
```

Then we call the `nginx/kustomization` module.

```hcl
module "nginx_ingress" {
  source = "kbst.xyz/catalog/nginx/kustomization"
  version = "0.46.0-kbst.1"

  configuration_base_key = "default"
  configuration = {
    default = {
      replicas = [{
        name = "ingress-nginx-controller"
        count = 2
      }]
    }
  }
}
```

Unlike for the Helm based module though, running `terraform plan` now will individually show all Kubernetes resources and their actual configuration.

> To keep this blogpost palatable, I only show the details for the namespace.

Reviewer: ```suggestion
To keep this blog post palatable, I show the details for the namespace only.

```

module.nginx_ingress.kustomization_resource.p1["apps_v1_Deployment|ingress-nginx|ingress-nginx-controller"]: Creating...
module.nginx_ingress.kustomization_resource.p1["rbac.authorization.k8s.io_v1_Role|ingress-nginx|ingress-nginx-admission"]: Creating...
module.nginx_ingress.kustomization_resource.p1["~G_v1_ServiceAccount|ingress-nginx|ingress-nginx-admission"]: Creating...
module.nginx_ingress.kustomization_resource.p1["rbac.authorization.k8s.io_v1_RoleBinding|ingress-nginx|ingress-nginx"]: Creating...

...

module.nginx_ingress.kustomization_resource.p1["batch_v1_Job|ingress-nginx|ingress-nginx-admission-create"]: Creation complete after 3s [id=a12aea4c-c92c-4655-912e-e8a0ca9bb67a]
module.nginx_ingress.kustomization_resource.p2["admissionregistration.k8s.io_v1_ValidatingWebhookConfiguration|~X|ingress-nginx-admission"]: Creating...
module.nginx_ingress.kustomization_resource.p2["admissionregistration.k8s.io_v1_ValidatingWebhookConfiguration|~X|ingress-nginx-admission"]: Creation complete after 0s [id=b1046e97-bfa7-4427-96d7-c83137da2f8a]

Apply complete! Resources: 18 added, 0 changed, 0 destroyed.
```

> Naturally this also means, that if we increase the replica count here.

Reviewer: ```suggestion
Naturally, this also means, that if we increase the replica count here.

```

```hcl
replicas = [{
  name = "ingress-nginx-controller"
  count = 3
}]
```

> The `terraform plan` will show which Kubernetes resource will change and what the diff is.

Reviewer: ```suggestion
...the `terraform plan` shows which Kubernetes resources will change and what the diff is.

```

```hcl
Terraform will perform the following actions:

  # module.nginx_ingress.kustomization_resource.p1["apps_v1_Deployment|ingress-nginx|ingress-nginx-controller"] will be updated in-place
  ~ resource "kustomization_resource" "p1" {
        id       = "366bd10d-745e-4172-9153-35ff5fad53d8"
      ~ manifest = jsonencode(
          ~ {
              ~ spec       = {
                  ~ replicas             = 2 -> 3
                    # (4 unchanged elements hidden)
                }
                # (3 unchanged elements hidden)
            }
        )
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Maybe more importantly even, the Kustomization provider will also correctly show if a resource can be changed using an in-place update.

> Or if a destroy and recreate is required because there is a change to an immutable field for example.

Reviewer: ```suggestion
Or if a destroy and recreate is required because there is a change to an immutable field, for example.

```

> Or if a destroy-and-recreate is required because there is a change to an immutable field, for example.

Reviewer: For non-experts like myself, would you explicitly identify the items that support this important point?

Author: Good point, I've added a section that explains the why and shows how this looks in the Terraform plan now.

## Try it yourself

> If you want to try the modules yourself, you can either use one of the modules from the catalog that bundle upstream YAML, like the [Prometheus operator](https://www.kubestack.com/catalog/prometheus), [Cert-Manager](https://www.kubestack.com/catalog/cert-manager), [Sealed secrets](https://www.kubestack.com/catalog/sealed-secrets) or [Tekton](https://www.kubestack.com/catalog/tektoncd) for example.

Reviewer: ```suggestion
If you want to try the modules yourself, you can either use one of the modules from the catalog that bundle upstream YAML, like the [Prometheus operator](https://www.kubestack.com/catalog/prometheus), [Cert-Manager](https://www.kubestack.com/catalog/cert-manager), [Sealed secrets](https://www.kubestack.com/catalog/sealed-secrets) or [Tekton](https://www.kubestack.com/catalog/tektoncd), for example.

```
